pkg/shortnames/shortnames.go

```diff
@@ -0,0 +1,472 @@
+package shortnames
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/pkg/sysregistriesv2"
+	"github.com/containers/image/v5/types"
+	"github.com/manifoldco/promptui"
+	"github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+// IsShortName returns true if the specified input is a "short name" and an
+// error that may occur when the input could not be parsed.  A "short name"
+// refers to a container image without a fully-qualified reference, and is
+// hence missing a registry (or domain).  Names including a digest are not
+// short names.
+//
+// Examples:
+//  * short names: "image:tag", "library/fedora"
+//  * not short names: "quay.io/image", "localhost/image:tag",
+//                     "server.org:5000/lib/image", "image@sha256:..."
+func IsShortName(input string) (bool, error) {
+	isShort, _, err := isShortName(input)
+	return isShort, err
+}
+
+// isShortName implements IsShortName but also returns the parsed
+// reference.Reference.
+func isShortName(input string) (bool, reference.Reference, error) {
+	ref, err := reference.Parse(input)
+	if err != nil {
+		return false, ref, errors.Wrapf(err, "cannot parse input: %q", input)
+	}
+
+	named, ok := ref.(reference.Named)
+	if !ok {
+		return true, ref, nil
+	}
+
+	registry := reference.Domain(named)
+	if strings.ContainsAny(registry, ".:") || registry == "localhost" {
+		return false, ref, nil
+	}
+
+	return true, ref, nil
+
+}
+
+// normalizeUserInput normalizes the user-specified reference.  Namely, it
+// strips off the tag or digest and stores it in the return values so that both
+// can be readded to possible resolved alias' value at a later point.
+func normalizeUserInput(ref reference.Reference) (isTagged bool, isDigested bool, normalized string, tag string, digest digest.Digest) {
+	normalized = ref.String()
+
+	named, isNamed := ref.(reference.Named)
+	if !isNamed {
+		return
+	}
+
+	// Strip off tag.
+	tagged, isT := named.(reference.NamedTagged)
+	if isT {
+		isTagged = true
+		tag = tagged.Tag()
+		normalized = reference.TrimNamed(tagged).String()
+		return
+	}
+
+	// Strip off digest.
+	digested, isD := named.(reference.Digested)
+	if isD {
+		isDigested = true
+		digest = digested.Digest()
+		normalized = reference.TrimNamed(named).String()
+		return
+	}
+
+	return
+}
+
+// GuessPullError returns an error that _guesses_ why a the specified short
+// name could not resolve to a successful pull candidate.  This can be helpful
+// for giving some hints to users to resolve pull errors.  Note that nil is
+// returned if it's not a short name.
+//
+// GuessPullError is expected to be called if none of the `Resolved` candidates
+// from `Resolve` lead to a successfully pulled image.
+func GuessPullError(ctx *types.SystemContext, name string) error {
```

Comment:
@mtrmac, this could be a tricky one. @rhatdan asked to improve the errors and I think Buildah and Podman should share this code.

@rhatdan PTAL

Reply:
Now further generalized in `func GuessPullError(ctx *types.SystemContext, name string, pullErrors []error) error {`